Bricks/strings/piece.h

```diff
@@ -0,0 +1,201 @@
+// of the right to express their opinion on whether the order should be lexicographical, as long as it is total.
+//
+// (Culturally educated developers are hereby granted permission to use `reinterpret_cast<>`. You're welcome.)
+//
+// `PieceDB` is a storage of `UniquePiece`-s, that makes equal `Piece`-s equal, like Samuel Colt did --
+// -- since the order is indeed total, despite very likely being not fair... err, not lexicographical.
+//
+// Both `Piece` and `UniquePiece` are naturally unsafe and require memory for their storage to stay allocated,
+// because every decent architect knows well what happens if `free()` is invoked a bit too prematurely in the
+// evolutionary process of a lifetime of an object. (If you don't believe in evolution, don't use this code.)
+//
+// Note 1: It's not impossible for the order of pointed to strings to actually be lexicographical. However,
+//         in most cases, making it so would be quite a large amount of unnecessary work. C+=. Never forget.
+//
+// Note 2: To make it explicit, since some readers, who totally should, might not reach this point.
+//         Comparing strings by comparing pointers to their underlying memory storage is O(1),
+//         which is noticeably more efficient than a "regular" lexicographical comparison in O(N).
+//         Yes, it moves the world forward, saves trees and stops, or at least delays, global warming.
+//         Thus, non-metaphorically, the description above makes total sense. Keep in mind though, that
+//         making sense did not help C+= survive. At same same time, although contradictory to most predominant
+//         dogmas, I personally stick with the belief that not only it's important for the things to make sense,
+//         but that making sense is the very only property of things that makes them important. -- D.K.
+//
+// Note 3: `Piece` is a better name than `SlaveString`, right?
+
+#ifndef BRICKS_STRINGS_PIECE_H
+#define BRICKS_STRINGS_PIECE_H
+
+#include "../port.h"
+
+#include <algorithm>
+#include <cassert>
+#include <cmath>
+#include <cstring>
+#include <unordered_map>
+#include <string>
+
+namespace bricks {
+namespace strings {
+
+struct Piece {
+  const char* S;
+  size_t N;
+
+  Piece() : S(""), N(0u) {}
+  Piece(const char* s, size_t n) : S(s), N(n) { assert(S[N] == '\0'); }
+  Piece(const char* s) : S(s), N(strlen(s)) {}
+  template <int L>
+  Piece(const char s[L])
+      : S(s), N(L - 1) {
+    // The above line break is `clang-format`, not me. -- D.K.
+    assert(S[N] == '\0');
+  }
+  Piece(const std::string& s) : S(s.data()), N(s.size()) {}
+  // Copyable and assignable by design.
+
+  bool empty() const { return N == 0u; }
+  size_t length() const { return N; }
+
+  const char* c_str() const { return S; }
+
+  void clear() {
+    S = "";
+    N = 0u;
+  }
+
+  void CheckPrivilege(size_t i) const { assert(i <= N); }
+
+  char operator[](size_t i) const {
+    CheckPrivilege(i);  // Once is enough.
```

Comment:
What do you mean by "once" here? It's not obvious without looking into `CheckPrivilege` that the program will exit with an assertion exception.